### PR TITLE
Abort fork on intermediate states

### PR DESCRIPTION
**Pull request by AviAvni**

### CODE REVIEW

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/commands/cmd_debug.c [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+void ModuleEventHandler_AUXAfterKeyspaceEvent(void);
+
+// Global array tracking all extant GraphContexts (defined in module.c)
+extern GraphContext **graphs_in_keyspace;
+extern uint aux_field_counter;
+
+int Graph_Debug(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+	ASSERT(ctx != NULL);
+	// ASSERT(graphs_in_keyspace != NULL);
+	RedisModule_ReplicateVerbatim(ctx);
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/commands/cmd_debug.c:22]: yes so aux start will be called on slave

File: src/serializers/graphcontext_type.c [hunk 2]
```diff
@@ -12,6 +12,8 @@
 #include "decoders/decode_previous.h"
 #include "../util/redis_version.h"
 
+extern uint aux_field_counter;
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/serializers/graphcontext_type.c:15]: need it to test the aux counter

File: src/graph/graphcontext.c [hunk 3]
```diff
@@ -34,7 +33,7 @@ static inline void _GraphContext_IncreaseRefCount(GraphContext *gc) {
 
 static inline void _GraphContext_DecreaseRefCount(GraphContext *gc) {
 	// If the reference count is less than 0, the graph has been marked for deletion and no queries are active - free the graph.
-	if(__atomic_sub_fetch(&gc->ref_count, 1, __ATOMIC_RELAXED) < 0) {
+	if(__atomic_sub_fetch(&gc->ref_count, 1, __ATOMIC_RELAXED) == -1) {
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/graph/graphcontext.c:36]: before or after?

File: src/module_event_handlers.c [hunk 4]
```diff
@@ -188,11 +180,14 @@ static void _ClearKeySpaceMetaKeys(RedisModuleCtx *ctx, bool decode) {
...
-		// If a flushall occurs during replication, stop all decoding.
+	// reset `aux_field_counter` upon handeling FLUSH-ALL
+	if(eid.id == REDISMODULE_EVENT_FLUSHDB &&
+       subevent == REDISMODULE_SUBEVENT_FLUSHDB_START) {
 		aux_field_counter = 0;
-		currently_decoding_graphs = 0;
-		_ResetDecodeStates();
+		uint count = array_len(graphs_in_keyspace);
+		for (size_t i = 0; i < count; i++) {
+			GraphContext_Delete(graphs_in_keyspace[i]);
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/module_event_handlers.c:189]: in that case when redis call the free function it supply the freed graph context
there is nothing to remove from registry and the refcount going to be decreased to -2 and to not free the graph context

File: src/graph/graphcontext.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (swilly22) [reply src/graph/graphcontext.c:36]: After

File: src/module_event_handlers.c [hunk 5]
```diff
@@ -213,11 +208,31 @@ static bool _IsEventPersistenceEnd(RedisModuleEvent eid, uint64_t subevent) {
 		   );
 }
 
-// Server persistence event handler.
+// server persistence event handler
 static void _PersistenceEventHandler(RedisModuleCtx *ctx, RedisModuleEvent eid,
 		uint64_t subevent, void *data) {
-	if(_IsEventPersistenceStart(eid, subevent)) _CreateKeySpaceMetaKeys(ctx);
-	else if(_IsEventPersistenceEnd(eid, subevent)) _ClearKeySpaceMetaKeys(ctx, false);
+	if(INTERMEDIATE_GRAPHS) {
```
[the remark this answers is not in the document]
Comment (swilly22) [reply src/module_event_handlers.c:214]: Correct, but this seems more clear / make sense to me if we move it.

File: src/serializers/graphcontext_type.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (swilly22) [reply src/serializers/graphcontext_type.c:15]: there's no use for `aux_field_counter` in this file

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/module_event_handlers.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/module_event_handlers.c:214]: then this function will look like 
```
// server persistence event handler
static void _PersistenceEventHandler(RedisModuleCtx *ctx, RedisModuleEvent eid,
		uint64_t subevent, void *data) {
	if(INTERMEDIATE_GRAPHS && !process_is_child) {
		// don't mess with the keyspace if we have half-baked graphs
		return;
	}

	if(_IsEventPersistenceStart(eid, subevent)) {
		if(INTERMEDIATE_GRAPHS && process_is_child) {
		// check for half-baked graphs
		// indicated by `aux_field_counter` > 0
		// in such case we do not want to either perform backup nor do we want to
		// synchronize our replica, as such we're aborting by existing
		// assuming we're running on a fork process
		if(process_is_child) {
			// intermediate graph(s) detected, exit!
			RedisModule_Log(NULL, REDISMODULE_LOGLEVEL_WARNING,
					"RedisGraph - aborting BGSAVE, detected intermediate graph(s)");

			exit(255);
		}
	}
		_CreateKeySpaceMetaKeys(ctx);
	} else if(_IsEventPersistenceEnd(eid, subevent)) {
		_ClearKeySpaceMetaKeys(ctx, false);
	}
}
```

File: src/graph/graphcontext.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/graph/graphcontext.c:36]: I don't think so in our case we can get to -2 but want to call free once

File: src/commands/cmd_debug.c [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+#include "../module_event_handlers.h"
+
+void ModuleEventHandler_AUXBeforeKeyspaceEvent(void);
+void ModuleEventHandler_AUXAfterKeyspaceEvent(void);
+
+// Global array tracking all extant GraphContexts (defined in module.c)
+extern GraphContext **graphs_in_keyspace;
+extern uint aux_field_counter;
+
+static void Debug_AUX(const char *arg) {
```
Comment (swilly22) [inline src/commands/cmd_debug.c:19]: pass original `argv` and `argc` you might want to pass `argv+1, argc-1` as the first value `AUX` was consumed.

File: tests/flow/test_replication_states.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+                self._step(scenario[0], 1, None)
+
+                self._connection_permutation(connection_permutation, 2)
+
+                self._step(scenario[1], 2, None)
+
+                self._connection_permutation(connection_permutation, 3)
+
+                self._step(scenario[2], 3, None)
```
Comment (swilly22) [inline tests/flow/test_replication_states.py:114]: remove last argument, always `None`

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/commands/cmd_debug.c [hunk 8]
```diff
@@ -16,7 +16,11 @@ void ModuleEventHandler_AUXAfterKeyspaceEvent(void);
 extern GraphContext **graphs_in_keyspace;
 extern uint aux_field_counter;
 
-static void Debug_AUX(const char *arg) {
+static void Debug_AUX(RedisModuleString **argv, int argc) {
+	if(argc < 2) return;
```
Comment (swilly22) [inline src/commands/cmd_debug.c:20]: @AviAvni Emit an error in this case

**Review by swilly22 — APPROVED** [review]

**Comment by swilly22** [comment]
@AviAvni Cherry pick into 2.4

**Pull request merged**